File: VB/MyListServert.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Data
Imports System.Collections.Generic
Imports DevExpress.Data.Filtering
Imports System.ComponentModel

Namespace DXGrid_EF4_ServerMode
	Public Class MyListServert
		Implements IListServer
		Private list As IListServer
		Private summaryReplacements As IDictionary(Of String, String)
		Public Sub New(ByVal list As IListServer, ByVal summaryReplacements As IDictionary(Of String, String))
			Me.list = list
			Me.summaryReplacements = summaryReplacements
		End Sub

        'Private Sub list_ExceptionThrown(ByVal sender As Object, ByVal e As ServerModeExceptionThrownEventArgs)

        'End Sub

		#Region "IListServer Members"


        Public Event ExceptionThrown As EventHandler(Of ServerModeExceptionThrownEventArgs) Implements IListServer.ExceptionThrown
        Public Event InconsistencyDetected As EventHandler(Of ServerModeInconsistencyDetectedEventArgs) Implements IListServer.InconsistencyDetected

		Private Sub Apply(ByVal filterCriteria As DevExpress.Data.Filtering.CriteriaOperator, ByVal sortInfo As ICollection(Of ServerModeOrderDescriptor), ByVal groupCount As Integer, ByVal groupSummaryInfo As ICollection(Of ServerModeSummaryDescriptor), ByVal totalSummaryInfo As ICollection(Of ServerModeSummaryDescriptor)) Implements IListServer.Apply
			Dim newSortInfo As New List(Of ServerModeOrderDescriptor)()
			If sortInfo IsNot Nothing Then
				For Each item As ServerModeOrderDescriptor In sortInfo
					Dim newItem As ServerModeOrderDescriptor = item
					If TypeOf item.SortExpression Is OperandProperty AndAlso summaryReplacements.ContainsKey((CType(item.SortExpression, OperandProperty)).PropertyName) Then
						newItem = New ServerModeOrderDescriptor(New OperandProperty(summaryReplacements((CType(item.SortExpression, OperandProperty)).PropertyName)), item.IsDesc)
					End If
					newSortInfo.Add(newItem)
				Next item
			End If
			list.Apply(filterCriteria, newSortInfo, groupCount, groupSummaryInfo, totalSummaryInfo)
		End Sub

        Private Function FindIncremental(ByVal expression As DevExpress.Data.Filtering.CriteriaOperator, ByVal value As String, ByVal startIndex As Integer, ByVal searchUp As Boolean, ByVal ignoreStartRow As Boolean, ByVal allowLoop As Boolean) As Integer Implements IListServer.FindIncremental
            Return list.FindIncremental(expression, value, startIndex, searchUp, ignoreStartRow, allowLoop)
        End Function

        Private Function GetAllFilteredAndSortedRows() As System.Collections.IList Implements IListServer.GetAllFilteredAndSortedRows
            Return list.GetAllFilteredAndSortedRows()
        End Function

        Private Function GetGroupInfo(ByVal parentGroup As ListSourceGroupInfo) As List(Of ListSourceGroupInfo) Implements IListServer.GetGroupInfo
            Return list.GetGroupInfo(parentGroup)
        End Function

        Private Function GetRowIndexByKey(ByVal key As Object) As Integer Implements IListServer.GetRowIndexByKey
            Return list.GetRowIndexByKey(key)
        End Function

        Private Function GetRowKey(ByVal index As Integer) As Object Implements IListServer.GetRowKey
            Return list.GetRowKey(index)
        End Function

        Private Function GetTotalSummary() As Dictionary(Of Object, Object) Implements IListServer.GetTotalSummary
            Return list.GetTotalSummary()
        End Function

        Private Function GetUniqueColumnValues(ByVal expression As DevExpress.Data.Filtering.CriteriaOperator, ByVal maxCount As Integer, ByVal includeFilteredOut As Boolean) As Object() Implements IListServer.GetUniqueColumnValues
            Return list.GetUniqueColumnValues(expression, maxCount, includeFilteredOut)
        End Function


        Private Function LocateByValue(ByVal expression As DevExpress.Data.Filtering.CriteriaOperator, ByVal value As Object, ByVal startIndex As Integer, ByVal searchUp As Boolean) As Integer Implements IListServer.LocateByValue
            Return list.LocateByValue(expression, value, startIndex, searchUp)
        End Function

        Private Sub Refresh() Implements IListServer.Refresh
            list.Refresh()
        End Sub

#End Region

#Region "IList Members"

        Private Function Add(ByVal value As Object) As Integer Implements System.Collections.IList.Add
            Return list.Add(value)
        End Function

        Private Sub Clear() Implements System.Collections.IList.Clear
            list.Clear()
        End Sub

        Private Function Contains(ByVal value As Object) As Boolean Implements System.Collections.IList.Contains
            Return list.Contains(value)
        End Function

        Private Function IndexOf(ByVal value As Object) As Integer Implements System.Collections.IList.IndexOf
            Return list.IndexOf(value)
        End Function

        Private Sub Insert(ByVal index As Integer, ByVal value As Object) Implements System.Collections.IList.Insert
            list.Insert(index, value)
        End Sub

        Private ReadOnly Property IsFixedSize() As Boolean Implements System.Collections.IList.IsFixedSize
            Get
                Return list.IsFixedSize
            End Get
        End Property

        Private ReadOnly Property IsReadOnly() As Boolean Implements System.Collections.IList.IsReadOnly
            Get
                Return list.IsReadOnly
            End Get
        End Property

        Private Sub Remove(ByVal value As Object) Implements System.Collections.IList.Remove
            list.Remove(value)
        End Sub

        Private Sub RemoveAt(ByVal index As Integer) Implements System.Collections.IList.RemoveAt
            list.RemoveAt(index)
        End Sub

        Public Property IList_Item(ByVal index As Integer) As Object Implements System.Collections.IList.Item
            Get
                Return list(index)
            End Get
            Set(ByVal value As Object)
                list(index) = value
            End Set
        End Property

#End Region

#Region "ICollection Members"

        Private Sub CopyTo(ByVal array As Array, ByVal index As Integer) Implements System.Collections.ICollection.CopyTo
            list.CopyTo(array, index)
        End Sub

        Private ReadOnly Property Count() As Integer Implements System.Collections.ICollection.Count
            Get
                Return list.Count
            End Get
        End Property

        Private ReadOnly Property IsSynchronized() As Boolean Implements System.Collections.ICollection.IsSynchronized
            Get
                Return list.IsSynchronized
            End Get
        End Property

        Private ReadOnly Property SyncRoot() As Object Implements System.Collections.ICollection.SyncRoot
            Get
                Return list.SyncRoot
            End Get
        End Property

#End Region

#Region "IEnumerable Members"

        Private Function GetEnumerator() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
            Return list.GetEnumerator()
        End Function

#End Region
    End Class
	Public Class MyListSource
		Implements IListSource
		Private source As IListSource
		Private summaryReplacements As IDictionary(Of String, String)
		Public Sub New(ByVal source As IListSource, ByVal summaryReplacements As IDictionary(Of String, String))
			Me.source = source
			Me.summaryReplacements = summaryReplacements
		End Sub


		#Region "IListSource Members"

		Private ReadOnly Property ContainsListCollection() As Boolean Implements IListSource.ContainsListCollection
			Get
				Return False
			End Get
		End Property

		Private Function GetList() As System.Collections.IList Implements IListSource.GetList
			Return New MyListServert(CType(source.GetList(), IListServer), summaryReplacements)
		End Function

		#End Region
	End Class
End Namespace

File: CS/MyListServert.cs
using System;
using DevExpress.Data;
using System.Collections.Generic;
using DevExpress.Data.Filtering;
using System.ComponentModel;

namespace DXGrid_EF4_ServerMode
{
    public class MyListServert : IListServer
    {
        IListServer list;
        IDictionary<string, string> summaryReplacements;
        public MyListServert(IListServer list, IDictionary<string, string> summaryReplacements)
        {
            this.list = list;
            this.summaryReplacements = summaryReplacements;
        }

        void list_ExceptionThrown(object sender, ServerModeExceptionThrownEventArgs e)
        {

        }

        #region IListServer Members

        void IListServer.Apply(DevExpress.Data.Filtering.CriteriaOperator filterCriteria, ICollection<ServerModeOrderDescriptor> sortInfo, int groupCount, ICollection<ServerModeSummaryDescriptor> groupSummaryInfo, ICollection<ServerModeSummaryDescriptor> totalSummaryInfo)
        {
            List<ServerModeOrderDescriptor> newSortInfo = new List<ServerModeOrderDescriptor>();
            if (sortInfo != null)
            {
                foreach (ServerModeOrderDescriptor item in sortInfo)
                {
                    ServerModeOrderDescriptor newItem = item;
                    if (item.SortExpression is OperandProperty && summaryReplacements.ContainsKey(((OperandProperty)item.SortExpression).PropertyName))
                    {
                        newItem = new ServerModeOrderDescriptor(new OperandProperty(summaryReplacements[((OperandProperty)item.SortExpression).PropertyName]), item.IsDesc);
                    }
                    newSortInfo.Add(newItem);
                }
            }
            list.Apply(filterCriteria, newSortInfo, groupCount, groupSummaryInfo, totalSummaryInfo);
        }

        event EventHandler<ServerModeExceptionThrownEventArgs> IListServer.ExceptionThrown
        {
            add { list.ExceptionThrown += value; }
            remove { list.ExceptionThrown -= value; }
        }

        int IListServer.FindIncremental(DevExpress.Data.Filtering.CriteriaOperator expression, string value, int startIndex, bool searchUp, bool ignoreStartRow, bool allowLoop)
        {
            return list.FindIncremental(expression, value, startIndex, searchUp, ignoreStartRow, allowLoop);
        }

        System.Collections.IList IListServer.GetAllFilteredAndSortedRows()
        {
            return list.GetAllFilteredAndSortedRows();
        }

        List<ListSourceGroupInfo> IListServer.GetGroupInfo(ListSourceGroupInfo parentGroup)
        {
            return list.GetGroupInfo(parentGroup);
        }

        int IListServer.GetRowIndexByKey(object key)
        {
            return list.GetRowIndexByKey(key);
        }

        object IListServer.GetRowKey(int index)
        {
            return list.GetRowKey(index);
        }

        Dictionary<object, object> IListServer.GetTotalSummary()
        {
            return list.GetTotalSummary();
        }

        object[] IListServer.GetUniqueColumnValues(DevExpress.Data.Filtering.CriteriaOperator expression, int maxCount, bool includeFilteredOut)
        {
            return list.GetUniqueColumnValues(expression, maxCount, includeFilteredOut);
        }

        event EventHandler<ServerModeInconsistencyDetectedEventArgs> IListServer.InconsistencyDetected
        {
            add { list.InconsistencyDetected += value; }
            remove { list.InconsistencyDetected -= value; }
        }

        int IListServer.LocateByValue(DevExpress.Data.Filtering.CriteriaOperator expression, object value, int startIndex, bool searchUp)
        {
            return list.LocateByValue(expression, value, startIndex, searchUp);
        }

        void IListServer.Refresh()
        {
            list.Refresh();
        }

        #endregion

        #region IList Members

        int System.Collections.IList.Add(object value)
        {
            return list.Add(value);
        }

        void System.Collections.IList.Clear()
        {
            list.Clear();
        }

        bool System.Collections.IList.Contains(object value)
        {
            return list.Contains(value);
        }

        int System.Collections.IList.IndexOf(object value)
        {
            return list.IndexOf(value);
        }

        void System.Collections.IList.Insert(int index, object value)
        {
            list.Insert(index, value);
        }

        bool System.Collections.IList.IsFixedSize
        {
            get { return list.IsFixedSize; }
        }

        bool System.Collections.IList.IsReadOnly
        {
            get { return list.IsReadOnly; }
        }

        void System.Collections.IList.Remove(object value)
        {
            list.Remove(value);
        }

        void System.Collections.IList.RemoveAt(int index)
        {
            list.RemoveAt(index);
        }

        object System.Collections.IList.this[int index]
        {
            get
            {
                return list[index];
            }
            set
            {
                list[index] = value;
            }
        }

        #endregion

        #region ICollection Members

        void System.Collections.ICollection.CopyTo(Array array, int index)
        {
            list.CopyTo(array, index);
        }

        int System.Collections.ICollection.Count
        {
            get { return list.Count; }
        }

        bool System.Collections.ICollection.IsSynchronized
        {
            get { return list.IsSynchronized; }
        }

        object System.Collections.ICollection.SyncRoot
        {
            get { return list.SyncRoot; }
        }

        #endregion

        #region IEnumerable Members

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return list.GetEnumerator();
        }

        #endregion
    }
    public class MyListSource : IListSource
    {
        IListSource source;
        IDictionary<string, string> summaryReplacements;
        public MyListSource(IListSource source, IDictionary<string, string> summaryReplacements)
        {
            this.source = source;
            this.summaryReplacements = summaryReplacements;
        }


        #region IListSource Members

        bool IListSource.ContainsListCollection
        {
            get { return false; }
        }

        System.Collections.IList IListSource.GetList()
        {
            return new MyListServert((IListServer)source.GetList(), summaryReplacements);
        }

        #endregion
    }
}
